storage/file_persister.go

```diff
@@ -13,14 +13,91 @@ import (
 	"path/filepath"
 	"strings"
 	"time"
+
+	"github.com/grafana/xk6-browser/env"
 )
 
+// FilePersister is the type that all file persisters must implement. It's job is
+// to persist a file somewhere, hiding the details of where and how from the caller.
+type FilePersister interface {
+	Persist(ctx context.Context, path string, data io.Reader) (err error)
+}
```

Comment:
I'm not sure about where this should live. It feels correct, since this is needed elsewhere in the codebase, but would it be better to define the interface multiple times closer to where it's needed?

I took the inspiration of this design from how `context` is implemented, and how the interface and concrete types seem to live in the same package, and so not having to continually write an interface over and over again.

Reply:
> but would it be better to define the interface multiple times

Where do we need to define it multiple times? AFAICS, there is only one usage in the `common` package, no?

Reply:
> Where do we need to define it multiple times? AFAICS, there is only one usage in the common package, no?

In the `browser` package in `module.go` and `moduelVU.go`

Reply:
> This is an anti-pattern

👍 

> The interface should live in the common package since its user is the common package

And in the `browser` package.

> This way, changing the methods on the interface won't break other packages and produce a cascading effect of maintenance burden.

👍 

> and concrete types are never exported, unlike we did

What if we unexported the concrete types? Would that be acceptable?

Reply:
The `Context` interface is exported because it wants others to implement this interface. It's a standard library package, after all. In our case, there are only two implementations, and they're in the same package. So, this is more like the Java/C# way of doing things with a factory pattern. Go doesn't need this.

Since Go has a structural type system, it's always OK to copy an interface in both packages if we don't expect others to implement the interface in different ways. The interface doesn't have to have a definite name and place, it doesn't have to be singular. The concrete types already provide documentation.

Moreover, by putting the factory in the `storage` package, we hide the decision to pick a storage. Since our code starts from the mapping layer, this makes code navigation difficult. We must go to the storage package to understand how it works.

So, I suggest:
- Adding an unexported interface in the `browser` package to make switching easier.
- Moving `NewFilePersister` into the mapping layer to make it explicit what concrete types it's passing to the inner layers.
- Adding an exported `Persister` interface in the `common` package so that the `common` package methods can accept the interface.

Reply:
Yep, the bit about the interfaces, concrete types, `context` and DRY  I agree with.

> Since our code starts from the mapping layer, this makes code navigation difficult

Why is this particularly an issue in this case? We have to navigate from file to file and package to package often, but it's not the greatest of burdens.

> Moving `NewFilePersister` into the mapping layer as an unexported function to make it explicit what concrete types it's passing to the inner layers

By doing this we're moving away from the factory pattern, and allowing the decision to be made by the mapping layer before it injects it into the appropriate APIs that require it, which would mean that the `NewFilePersister` would still need to return an interface. I'm not sure I understand the benefit of doing that. Can you explain why moving `NewFilePersister` to the mapping layer would benefit us?

> Adding an unexported interface in the browser package to make this package's switching between different types easier.

👍 

> Adding an exported Persister interface in the common package so that the common package methods can accept the interface.

👍

Reply:
> Why is this particularly an issue in this case?

Increases the cognitive load, especially when debugging. It's not about navigation per se but understanding and following the call path. A -> B -> C.

> which would mean that the NewFilePersister would still need to return an interface

Not necessarily. A method/function can read the environment variables and save the necessary concrete type to the module's interface field.

Reply:
I've moved the interfaces out into their own packages and unexported the types so that they can't be used outside of the package that they're defined in (7125509982597028318c97f70728dad4d952ae0c). I'm still unsure of the next step, which is to call `NewFilePersister` from the mapping layer or refactor/move `NewFilePersister` into the mapping layer?

> 1- Increases the cognitive load, especially when debugging.

👍 

> 2- We're hiding how we're picking a persister.

This is just abstracting that bit of detail away from the caller. I don't see why the `browser` or `common` packages need to know about how the `storage` package makes the decision.

> 3- We create a dependency between multiple packages, which is fragile and unnecessary

The only way around this is if the logic to make the decision to pick either the `LocalFilePersister` or the `RemoteFilePersister` resided in the `browser` package. It could then create either of the persisters, store it in the `moduleVU` within a package defined interface -- at this point should the `common` package export its version of the `filePersister` interface or not? Seems like a smell not to export it.

Reply:
The new code looks nicer but still needs improvement. So, it can be an example implementation to follow for the upcoming features other than screenshot persisting.

> Which is to call NewFilePersister from the mapping layer or refactor/move NewFilePersister into the mapping layer?

We're moving the driving logic that deals with externals in the `browser` package. I recommend moving the logic in `NewFilePersister` to the mapping layer and setting the `filePersister` field in the `browser` package for clarity and maintainability.

> This is just abstracting that bit of detail away from the caller. I don't see why the browser or common packages need to know about how the storage package makes the decision.

The caller is the `browser` package, which is the driver. It doesn't need this kind of abstraction since it's an abstraction already. It's not the best place to abstract the machinery. More cognitive load. For consistency and clarity, we should see what happens there and which concrete values it's passing to downstream components.

> The only way around this is if the logic to make the decision to pick either the LocalFilePersister or the RemoteFilePersister resided in the browser package. It could then create either of the persisters, store it in the moduleVU within a package defined interface. At this point the common package should export its own persister interface.

Yes, this will be great. An exported interface field in `filePersister`. It doesn't have to be package-defined. One of the methods in `RootModule`, or a function in the `browser` package can update this field.

> Or move all the persister code into the browser package

I would advise against this as we don't need to see the internal machinery related to persisting. We should only see the decision-making process of picking a persisted for clarity.

> if that's the case then I have the following questions:
> Why do we have a storage package?
> When should we create new packages?

Like I said above, `storage` package is perfectly fine to have. It should export those concrete persister types to let the `browser` package drive this package.

---

I have a few more notes about the tests.

Since we're abstracting the persistency logic in `common`, we should not call the `NewFilePersister` there. Instead, we can just pass `Screenshot`, a fake type that we implement in `tests`. This will remove the dependency of `storage` from `tests`.

Reply:
> We're moving the driving logic that deals with externals in the browser package. I recommend moving the logic in NewFilePersister to the mapping layer and setting the filePersister field in the browser package for clarity and maintainability.

I think i know what you are getting at now. The `browser` package's role is the entry point, initialiser of various global and vu level types, and a translation layer between goja and the internal business logic. We should rename and reword what the packages do, at the moment [it's not clear](https://github.com/grafana/xk6-browser/blob/205f1734dc72351c0ee9f609e46cfc494ac6f99e/browser/module.go#L1).

> The caller is the browser package, which is the driver. It doesn't need this kind of abstraction since it's an abstraction already. It's not the best place to abstract the machinery. More cognitive load. For consistency and clarity, we should see what happens there and which concrete values it's passing to downstream components.

Yep, I have a better understanding of what the `browser` package does now and why you are making the suggestion 👍 

> An unexported interface field in filePersister

Not sure what you mean here, is it not already that in the latest change? The only thing remaining is to move `NewFilePersister` out of `storage` and into `browser`. I'll make the change and we can discuss any issues in a new thread.

> I would advise against this as we don't need to see the internal machinery related to persisting. We should only see the decision-making process of picking a concrete persister for clarity.

The packages I see that are similar to `storage` are `chromium`, `trace`, `k6ext`, and `log`. I guess these aren't "drivers" but abstracted away implementation details, which makes them simpler to swap out for a new implementation. So these packages are less to do with reuse, and more to do with hiding the details? Maybe the docs for the packages need to be updated.

> Like I said above, storage package is perfectly fine to have. It should export those concrete persister types to let the browser package drive this package.

👍 

> Since we're abstracting the persistency logic in common, we should not call the NewFilePersister in the screenshot tests. Instead, we can just pass a fake type that we implement in tests to the screenshot tests. This will remove the dependency of storage from tests.

👍 I'll work on the tests once we've settled on the implementation.

Reply:
I'm moved the `newFilePersister` code in 33c426b08be03cbf3555f94fb56e033aaad0d047.

Reply:
> The browser package's role is the entry point

💯 We have an open issue about clarifying the role of the `browser` package.

https://github.com/grafana/xk6-browser/issues/1065

It's about renaming but can be extended or supplemented with other issues. I tackled this renaming work last month. But I couldn't pull it off because k6-core refers to the `browser` package, and renaming this package breaks k6-core. I couldn't find a good solution. We can, at least, update the package comment.

> We should rename and reword what the packages do, at the moment [it's not clear](https://github.com/grafana/xk6-browser/blob/205f1734dc72351c0ee9f609e46cfc494ac6f99e/browser/module.go#L1).

I'll do that after discussing it internally with you.

> Not sure what you mean here, is it not already that in the latest change?

Yes, it's in the latest commits. I just reiterated it. It's all fine.

> The packages I see that are similar to storage are chromium, trace, k6ext, and log. I guess these aren't "drivers" but abstracted away implementation details, which makes them simpler to swap out for a new implementation. So these packages are less to do with reuse, and more to do with hiding the details? Maybe the docs for the packages need to be updated.

👍 Yes, they're about doing their specific machinery and hiding the details of doing so. They're driven by other packages. I agree that we should update their docs.

> 👍 I'll work on the tests once we've settled on the implementation.

👍

Reply:
> I'm moved the newFilePersister code in https://github.com/grafana/xk6-browser/commit/33c426b08be03cbf3555f94fb56e033aaad0d047.

Thanks. This is excellent.